Change JAX type promotion to prefer inexact types.

hawkinsp opened the pull request:
NumPy's type promotion rules tend to promote aggressively to `float64`, which isn't a very accelerator-friendly behavior when not all accelerators (e.g., TPUs) support 64-bit floating point types. Even on accelerators that support 64-bit floating point types (e.g., GPUs), promotion to a 64-bit type comes with a significant performance cost.

This change makes JAX type promotion between inexact and exact types closer to PyTorch's promotion semantics, which are a better fit for modern accelerators:
e.g.,
```
import numpy as onp
from jax import numpy as np

In [1]: onp.promote_types(onp.float32, onp.int32)   
Out[1]: dtype('float64')

In [2]: onp.promote_types(onp.float16, onp.int64)   
Out[2]: dtype('float64')

In [3]: np.promote_types(onp.float32, onp.int32)    
Out[3]: dtype('float32')

In [4]: np.promote_types(onp.float16, onp.int64)    
Out[4]: dtype('float16')
```

This change is in preparation for enabling x64 mode by default on all platforms.

hawkinsp committed:
Change the JAX type promotion table to prefer inexact types during type promotion.

hawkinsp committed:
Document JAX type promotion.

hawkinsp committed:
Update type promotion docs.

hawkinsp committed:
Fix typo.

hawkinsp committed:
Fix sentence fragment.

shoyer commented on jax/numpy/lax_numpy.py:
```diff
@@ -2369,6 +2369,7 @@ def inner(a, b, precision=None):
 def outer(a, b, out=None):
   if out:
     raise NotImplementedError("The 'out' argument to outer is not supported.")
+  a, b = _promote_dtypes(a, b)
```
Is this really needed? I would guess that `*` below would handle this.

hawkinsp replied:
Yes, if we wish to have the same promotion behavior for `onp` and `np` arrays. My belief is that `jax.numpy` APIs should be consistent in their behavior regardless of which kind of array was passed to them.

Another option here would have been to add a `jit` decorator.

shoyer commented on jax/dtypes.py:
```diff
@@ -141,38 +141,72 @@ def issubdtype(a, b):
+  onp.dtype('uint8'),
+  onp.dtype('uint16'),
+  onp.dtype('uint32'),
+  onp.dtype('uint64'),
```
If it’s not too annoying, maybe flip the order of signed and unsigned integers? Then the dtypes kinds would be sorted by priority.

hawkinsp replied:
Done.

mattjj reviewed (approved):
LGTM! This is some of the final touches on a huge upgrade and our users are clambering for it. Thanks for taking the time to write detailed beautiful docs too. You set a high bar.

hawkinsp committed:
Flip the order of unsigned and signed types in the type promotion table.

Include the script that generated the type promotion table in the documentation.

shoyer commented:
awesome, thanks @hawkinsp !
